docs/DesignDocs/NotesOnAttestationAPI.md

```diff
@@ -0,0 +1,362 @@
+Notes on OE SDK Attestation API and Plugin Library Design
+====
+
+This document contains notes on the concepts and assumptions behind the design
+of the TEE-agnostic OE SDK Attestation API as described in document
+[Proposal of OE SDK Attestation Public and Plugin API for V0.10 Release](Attestation_API_Proposal.md).
+It also captures notes on the design of SGX plugins as explained in document
+[Attestation: OE SDK Integration with Intel® SGX SDK quote-ex Library for Generation of Evidence in New Formats](SGX_QuoteEx_Integration.md).
+
+Hopefully these notes can help readers' understanding of the API and its
+usage, and the SGX plugin design.
+
+# Notes on TEE-agnostic OE SDK Attestation API Design
+
+## Semantics of Custom Claims and Optional Parameters
+
+This section explains the custom claims vs optional parameters for the [OE SDK attestation API](Attestation_API_Proposal.md#proposed-oe-sdk-v010-public-api)
+function `oe_get_evidence()`.
+
+Among all the parameters for the OE SDK attestation API
+function `oe_get_evidence()`,
+two of them are a list of custom claims (`custom_claims`) and an optional
+parameters blob (`opt_params`). While in general the fields in these two
+parameters fit the definition of the term "Claims" defined in the
+[RATS document](https://tools.ietf.org/html/draft-ietf-rats-architecture),
+in the context of the function `oe_get_evidence()`, these two parameters
+have distinct semantics that might not be straightforward to application
+and plugin software developers.
+
+To help clarify what goes into the custom claims (`custom_claims`) and
+what goes into the optional parameters (`opt_params`) in the function
+`oe_get_evidence()`, it would be good to articulate the semantics of these
+two entities.
+
+- Custom claims have well-defined semantics:
+  - Custom claims can be any flat array of bytes supplied by attester
```

Comment:
I don't see the point of having plugin writers to implement their own serialization/deserialization routines for custom claims. The previous implementation defines an array of oe_claim_t objects, and provides serialization/deserialization for plugins. We should keep those, while at the same time, allow some plugins, to override the default serialization/deserialization routines if needed.

Reply:
@jxyang I wouldn't assume that two plugins serialize custom claims the same way.   The only value of having some common implementation is if it's common for many plugins to do it the same way.  Not sure we have enough evidence of that yet.   Many other formats are standard, not OE specific, so overriding would be the norm I would think.

Reply:
@dthaler  This is a pain point from an user point of view. I have personally used SGX plugin and EEID plugin. And I think they share the same set of custom claims. In fact, I am implementing attestation TLS for SGX plugin first and then move to EEID plugin later. Both of them depend on custom claims.